docs/design/proposals/pv-key-rotation.md

```diff
@@ -0,0 +1,55 @@
+# Encryption Key Rotation
+
+## Proposal
+
+Subject of this proposal is to add support for rotation of encryption keys (KEKs) for encrypted volumes in Ceph-CSI.
+
+## Document Terminology
+
+- Encryption Key: The passphrase that is used to encrypt and open the device.
+- LUKS: The specification used by dm-crypt to process encrypted volumes on linux.
+
+## Proposed Solution
+
+The proposed solution in this document, is to address the periodic rotation of encryption keys for encrypted volumes.
+
+This document outlines the rotation steps for PVCs backed by RBD and will be updated with other volume types as they are supported.
+
+### Implementation Summary
+
+This feature builds upon the foundation laid by encrypted pvcs.
+
+An existing storage class can be annotated with `keyrotation.csiaddons.openshift.io/schedule` to enable the key rotation. The value of this annotation can be schedule in cron format or one of the macros supported by K8s CronJob spec.
+
+The following new methods are added to `cryptsetup.go` for handling the key rotation.
+
+- `LuksAddKey`: Adds a new key to specified LUKS slot
+- `LuksRemoveKey`: Removes the specified key from its slot using `luksKillSlot`
+- `LuksVerifyKey`: Verifies that the given key exists in the given slot using `luksChangeKey`.
+
+### Implementation Details
+
+The encryption key rotation request will contain with it the volume ID, credentials and secrets.
+
+These values are then used to call `GenVolFromVolID` to get the rbdVolume structure.
+
+The `VolumeEncryption` struct is modified to make `generateNewEncryptionPassphrase` a public member function.
+
+A metadata is set on the RBD image to indicate that the image is being processed for keyrotation. Presence of this metadata will prevent the same image being processed again.
```

Comment:
What's the purpose of setting metadata and which server controller or node will set this / respond to this ? 
When will the rotation be triggered?

Reply:
We will register this service only on the node server hence the node server will set the metadata. Before doing that it will ensure that it is not already present.

The purpose of setting this metadata is to avoid multiple node servers trying to update the key at the same time.

The rotation will be triggered when the node server receives the RPC call from csi-addons. Which will be in turn be triggered by the cron schedule defined in the annotated storage class.

Reply:
The volume/operation locks which we have will serve that purpose ? 

I guess metadata is required if we have rbd mounted in multiple nodes but even with metadata set if controller pings another nodeplugin after first request then the status will get updated to complete?

Reply:
In case of RWX volumes we will send the request to only one of the nodes, as the underlying device is the same and doing it on one node will do it for all.

Ditching the metadata part then :)

Reply:
> In case of RWX volumes we will send the request to only one of the nodes, as the underlying device is the same and doing it on one node will do it for all.
> 
> Ditching the metadata part then :)

Yea, one more problem will be how to decide which node to target and send requests only to that.

Reply:
What I have so far is to get a list of vol attachments and see if the node advertises KeyRotate identity. If so, we send the request to it. We can safely ignore the rest of the nodes in the attachment list. This will work for both type of volumes as we need to find out the node to send rpc requests to anyways. WDYT?

Reply:
My main concern is request timing out and then we may choose another node in a follow up request. Operation locks cannot help in this case.

Reply:
what if we have a a recovery key generated by the controller and store it in the KMS temporarily until the existing key is updated in KMS? (we still need to check and see if this is a feasible and accepted solution)

Reply:
Just a thought, what about we requeue the reconcile and the next time we do not forward the request to the node that timed out? Leaning towards this as we won't have to change the existing KMS impl this way and we'll have much less processing on csi driver side of things.

Reply:
You need to think about the side effect it going to have multiple key or key not matching which is in KMS with the one actual in-use.